### PR TITLE
JDK-8357470: src/java.base/share/classes/sun/security/util/Debug.java implement the test for args.toLowerCase

### PR DESCRIPTION
**Pull request by myankelev**
* added an automated mixed case option
* using multithreading now
* added logs for simpler debug
* added missing -Djava.security.auth.debug coverage

### CODE REVIEW

File: test/jdk/sun/security/util/Debug/DebugOptions.java [hunk 1]
```diff
@@ -53,48 +57,170 @@ public class DebugOptions {
...
+                    return null;
+                });
+
+                System.out.println("option added to all tests " + Arrays.toString(params));
+            });
+
+            System.out.println("Starting all the threads");
+            final List<Future<Void>> res = executorService.invokeAll(testsCallables);
+            for (final Future<Void> future : res) {
+                future.get();
```
Comment (koushikthirupattur) [inline test/jdk/sun/security/util/Debug/DebugOptions.java:171]: future.get() could throw an unchecked exception. Can we consider logging which test failed and why?

File: test/jdk/sun/security/util/Debug/DebugOptions.java [hunk 2]
```diff
@@ -53,48 +57,170 @@ public class DebugOptions {
...
+                });
+                testsCallables.add(() -> {
+                    testMixedCaseParameter(
+                            "-Djava.security.auth.debug=",
+                            params[0],
+                            params[1],
+                            params[2]);
+                    return null;
+                });
+
```
Comment (koushikthirupattur) [inline test/jdk/sun/security/util/Debug/DebugOptions.java:164]: Can we add a test to explicitly test something like this for `testParameter("-Djava.security.debug=", "all", EXPECTED_ALL_REGEX, "not expected")` and `testMixedCaseParameter("-Djava.security.debug=", "AlL", EXPECTED_ALL_REGEX, "not expected")` ?

File: test/jdk/sun/security/util/Debug/DebugOptions.java [hunk 3]
```diff
@@ -53,48 +57,170 @@ public class DebugOptions {
...
+                System.out.println("option added to all tests " + Arrays.toString(params));
```
Comment (koushikthirupattur) [inline test/jdk/sun/security/util/Debug/DebugOptions.java]: Optional: consider labeling the log output to make test parameters easier to read when scanning test output

File: test/jdk/sun/security/util/Debug/DebugOptions.java [hunk 4]
```diff
@@ -53,48 +57,170 @@ public class DebugOptions {
...
+     * in both mixed and lowercase
+     */
+    @Test
+    public void debugOptionsTest() throws Exception {
+
+        try (final ExecutorService executorService = Executors.newVirtualThreadPerTaskExecutor()) {
+            final List<Callable<Void>> testsCallables = new ArrayList<>();
+
+            patternMatches.forEach(params -> {
+                testsCallables.add(() -> {
```
Comment (koushikthirupattur) [inline test/jdk/sun/security/util/Debug/DebugOptions.java:148]: Can we consider adding negative tests with invalid/malformed parameters like null, empty strings, unknown keywords, etc?

File: test/jdk/sun/security/util/Debug/DebugOptions.java [hunk 5]
```diff
@@ -53,48 +57,170 @@ public class DebugOptions {
...
+     * and last letters uppercase
+     * <p>
+     * e.g.:
+     * hello -> HellO
+     *
+     * @param paramString string to change
+     * @return resulting string
+     */
+    private String makeFirstAndLastLetterUppercase(final String paramString) {
+        final int length = paramString.length();
```
Comment (koushikthirupattur) [inline test/jdk/sun/security/util/Debug/DebugOptions.java:127]: Edge Case : Can we check for null or empty before this ?

File: test/jdk/sun/security/util/Debug/DebugOptions.java [hunk 6]
```diff
@@ -53,48 +57,170 @@ public class DebugOptions {
...
+     * This method will change the input string to have the first
+     * and last letters uppercase
+     * <p>
+     * e.g.:
+     * hello -> HellO
+     *
+     * @param paramString string to change
+     * @return resulting string
+     */
+    private String makeFirstAndLastLetterUppercase(final String paramString) {
```
Comment (koushikthirupattur) [inline test/jdk/sun/security/util/Debug/DebugOptions.java:124]: If the goal is to test case-insensitive parsing comprehensively, this method should be expanded to test various case combinations (e.g., all-uppercase, alternating case, etc). Otherwise, document this limitation.

Comment (myankelev) [reply test/jdk/sun/security/util/Debug/DebugOptions.java:124]: As per our discussion, split this test into 3. upper case, mixed case and lowercase. Done in the next commit

File: test/jdk/sun/security/util/Debug/DebugOptions.java (hunk 5 above)
Comment (myankelev) [reply test/jdk/sun/security/util/Debug/DebugOptions.java:127]: Done in the next commit

File: test/jdk/sun/security/util/Debug/DebugOptions.java (hunk 4 above)
Comment (myankelev) [reply test/jdk/sun/security/util/Debug/DebugOptions.java:148]: This is not in the scope of this change, however this will be addressed in [JDK-8360373](https://bugs.openjdk.org/browse/JDK-8360373)

File: test/jdk/sun/security/util/Debug/DebugOptions.java (hunk 2 above)
Comment (myankelev) [reply test/jdk/sun/security/util/Debug/DebugOptions.java:164]: As we discussed offline, if we only test a few options hardcoded like this we risk missing a potential bug where one option may not support mixed and upper case

File: test/jdk/sun/security/util/Debug/DebugOptions.java (hunk 1 above)
Comment (myankelev) [reply test/jdk/sun/security/util/Debug/DebugOptions.java:171]: It is logged in the `test*Parameter` methods when the action is executed. It would be saved to a separate log file by jtreg, one per thread

File: test/jdk/sun/security/util/Debug/DebugOptions.java (hunk 3 above)
Comment (myankelev) [reply test/jdk/sun/security/util/Debug/DebugOptions.java]: I'd prefer keeping this as is, as the parameters are also logged in the `test*Parameter` when the thread is actually executed. This is just to make sure that the parameters are added to the list of threads, which would be easier in this format, as they are set in this format in the list above. 

log: 
```
Option added to all lowercase tests [properties, properties\\[.*\\|main|\\d{4}-\\d{2}-\\d{2}.*\\]:, properties:]
```
code: 
```java 
new String[]{"properties",
             EXPECTED_PROP_REGEX,
             "properties:"},
```

**Comment by myankelev** [comment]
> As the original contributor of this test case, it's nice to see the speed up via use of ExecutorService.
> 
> The upper case and lower case property tests are somewhat questionable IMO. One or two lines of test code would suffice but now we have a lot of noise distracting from the original intent of the test code. One could argue that every system property in the JDK could be subject to the same uppercase/lower case testing but such tests don't exist. A few lines of test code can easily verify behavior. Anything else enters the fuzzy logic testing approach. Hope it's not a pattern that new tests have to adopt.

Do you think that keeping it just mixed case would be sufficient? I believe this approach might be too much for other tests, it seems to me like a lot for this one as well

**Comment by coffeys** [comment]
> > As the original contributor of this test case, it's nice to see the speed up via use of ExecutorService.
> > The upper case and lower case property tests are somewhat questionable IMO. One or two lines of test code would suffice but now we have a lot of noise distracting from the original intent of the test code. One could argue that every system property in the JDK could be subject to the same uppercase/lower case testing but such tests don't exist. A few lines of test code can easily verify behavior. Anything else enters the fuzzy logic testing approach. Hope it's not a pattern that new tests have to adopt.
> 
> Do you think that keeping it just mixed case would be sufficient? I believe this approach might be too much for other tests, it seems to me like a lot for this one as well

I'd welcome anything that helps reduce the extra code that's being added to test multiple case combinations for a system property. It strikes me as overkill.

**Comment by myankelev** [comment]
@koushikthirupattur @coffeys brought the cases to mixed case only in order to still test everything, but no going into the testing of the implementation of the `toLowerCase()`

File: test/jdk/sun/security/util/Debug/DebugOptions.java [hunk 7]
```diff
@@ -47,54 +51,132 @@ public class DebugOptions {
...
+
+        final OutputAnalyzer outputAnalyzer = ProcessTools.executeTestJava(
+                paramName + formattedParam,
+                "DebugOptions");
         outputAnalyzer.shouldHaveExitValue(0)
                 .shouldMatch(expected)
                 .shouldNotMatch(notExpected);
     }
 
+    /**
```
Comment (coffeys) [inline test/jdk/sun/security/util/Debug/DebugOptions.java:114]: approach looks fine - this comment could be removed or corrected

Comment (myankelev) [reply test/jdk/sun/security/util/Debug/DebugOptions.java:114]: Thanks! Done in the next commit

**Review by coffeys — APPROVED** [review]
